Rewrite of the specification of DNSSEC03 (#1189)
zonemaster/zonemaster: docs/public/specifications/tests/DNSSEC-TP/dnssec03.md

**Salt:** The salt parameter has been seen as a security feature but
[RFC 9276][RFC 9276#section-3.1], section 3.1, states that zones "SHOULD NOT use
a salt by indicating a zero-length salt value instead". The justification for
the recommendation is found in [RFC 9276][RFC 9276#section-2.4], section 2.4.

## Scope

This test case will just ignore non-responsive name servers or name servers not
giving a correct DNS response for an authoritative name server (covered by
[Connectivity01]).

This test case is only relevant if the zone has been DNSSEC signed.

## Inputs

* "Child Zone" - The domain name to be tested.
* "Public Suffix List Data" - The list or a copy of the list found at
  [Public Suffix List data].

## Summary

* If no DNSKEY records are found, no further investigation will be done.

Message Tag outputted              | Level   | Arguments| Message ID for message tag
:----------------------------------|:--------|:---------|:--------------------------------------------
DS03_ERR_MULT_NSEC3                | ERROR   | ns_list  | Multiple NSEC3 records when one is expected. Fetched from name servers "{ns_list}".

> DS03_ILLEGAL_HASH_ALGO             | ERRRO   | ns_list, algo_num | The following servers respond with an illegal hash algorithm for NSEC3 ({algo_num}). Fetched from name servers "{ns_list}".

Reviewer: The typo is still there.

> DS03_ILLEGAL_ITERATION_VALUE       | ERROR   | ns_list, int | The following servers respond with the NSEC3 iteration value {int} (not recommended). Fetched from name servers "{ns_list}".
> DS03_ILLEGAL_SALT_LENGTH           | WARNING | ns_list, int | The following servers respond with a non-empty salt in NSEC3 ({int} octets). Fetched from name servers "{ns_list}".

Reviewer: The current recommendation could maybe be made clearer in those messages. For example:
```suggestion
DS03_ILLEGAL_ITERATION_VALUE       | ERROR   | ns_list, int | The following servers respond with the NSEC3 iteration value {int}. The recommended practice is to set this value to 0. Fetched from name servers "{ns_list}".
DS03_ILLEGAL_SALT_LENGTH           | WARNING | ns_list, int | The following servers respond with a non-empty salt in NSEC3 ({int} octets). The recommended practice is to use an empty salt. Fetched from name servers "{ns_list}".
```

DS03_INCONSISTENT_HASH_ALGO        | ERROR   |         | Inconsistent hash algorithms in NSEC3 in responses for the child zone from different name servers.
DS03_INCONSISTENT_ITERATION        | ERROR   |         | Inconsistent NSEC3 iteration values in responses for the child zone from different name servers.
DS03_INCONSISTENT_NSEC3_FLAGS      | ERROR   |         | Inconsistent NSEC3 flags in responses for the child zone from different name servers.

> DS03_INCONSISTENT_SALT_LENGTH      | ERROR   |         | Inconsistent salt length in NSEC3 in responses for the child zone from different name servers.

Reviewer: ```suggestion
DS03_INCONSISTENT_SALT_LENGTH      | ERROR   |         | Inconsistent salt lengths in NSEC3 in responses for the child zone from different name servers.
```

Author: I would say singular since there could only be one length.

Reviewer: I don’t think so. We’re talking about several values, always two or more, that do not agree between each other. In the singular form, it implies that it’s self-contradictory (such as in “an inconsistent story”), which for a numeric value, makes less sense. Besides, we have inconsistent messages: the previous messages use “hash algorithms” and “iteration values” in their plural forms.

Author: I checked with a native English speaker with two of the messages, and he suggests singular.

DS03_LEGAL_EMPTY_SALT              | INFO    | ns_list | The following servers respond with a legal empty salt in NSEC3. Fetched from name servers "{ns_list}".
DS03_LEGAL_HASH_ALGO               | INFO    | ns_list | The following servers respond with a legal hash algorithm in NSEC3. Fetched from name servers "{ns_list}".
DS03_LEGAL_ITERATION_VALUE         | INFO    | ns_list | The following servers respond with no NSEC3 iterations (as recommended). Fetched from name servers "{ns_list}".
DS03_NO_DNSSEC_SUPPORT             | NOTICE  | ns_list | The zone is not DNSSEC signed or not properly DNSSEC signed. Testing for NSEC3 has been skipped. Fetched from name servers "{ns_list}".
DS03_NO_NSEC3                      | INFO    | ns_list | The zone does not use NSEC3. Testing for NSEC3 has been skipped. Fetched from name servers "{ns_list}".
DS03_NSEC3_OPT_OUT_DISABLED        | INFO    | ns_list | The following servers respond with NSEC3 opt-out disabled (as recommended). Fetched from name servers "{ns_list}".

> DS03_NSEC3_OPT_OUT_ENABLED_NON_TLD | NOTICE  | ns_list | The following servers respond with NSEC3 opt-out enabled (not recommended). Fetched from name servers "{ns_list}".

Reviewer: ```suggestion
DS03_NSEC3_OPT_OUT_ENABLED_NON_TLD | NOTICE  | ns_list | The following servers respond with NSEC3 opt-out enabled. The recommended practice is to disable opt-out. Fetched from name servers "{ns_list}".
```

DS03_SERVER_NO_NSEC3               | ERROR   | ns_list | The following name servers do not use NSEC3, but others do. Testing for NSEC3 has been skipped on the following servers. Fetched from name servers "{ns_list}".
DS03_UNASSIGNED_FLAG_USED          | ERROR   | ns_list, int | The following servers respond with an NSEC3 record where an unassigned flag is used (flag {int}). Fetched from name servers "{ns_list}".

The value in the Level column is the default severity level of the message. The
severity level can be changed in the [Zonemaster-Engine profile]. Also see the
[Severity Level Definitions] document.

The argument names in the Arguments column lists the arguments used in the
message. The argument names are defined in the [argument list].

The name server names are assumed to be available at the time when the msgid
is created, if the argument name is "ns" or "ns_list" even when in the
"[Test procedure]" below it is only referred to the IP address of the name
servers.

## Test procedure

In this section and unless otherwise specified below, the term "[DNSSEC Query]"

> follow the specification for DNS queries as specified in

Reviewer: ```suggestion
follows the specification for DNS queries as specified in
```

    5.  Name server IP address ("Multiple NSEC3").
    6.  Name server IP address and NSEC3 hash algorithm ("Hash Algorithm").
    7.  Name server IP address and NSEC3 flags ("NSEC3 Flags").
    8.  Name server IP address and NSEC3 iterations value ("NSEC3 Iterations").
    9.  Name server IP address and NSEC3 salt length ("NSEC3 Salt Length").

6.  For each name server IP address in *NS IP* do:

    1. Send *DNSKEY Query* to the name server IP.
    2. If at least one of the following criteria is met, then go to next name
       server IP:
       1. There is no DNS response.
       2. The [RCODE Name] in the response is not "NoError".
       3. The AA flag is not set in the response.
    3. If the response does not contain any DNSKEY record with owner name
       matching *Child Zone* in the answer section, add name server IP to the
       *Responds Without DNSKEY* set and go to next name server.
    4. Add name server IP to the *Responds With DNSKEY* set.
    5. Send *NSEC Query* to the name server IP and do:

       1. If the authority section contains no NSEC3 record then add the name
          server IP to the *Responds Without NSEC3* set and go to next name

>           server.

Reviewer: The question is, what should we do if we don’t get an authoritative NOERROR response on this one. In this case, can we just treat it as NODATA?

>           server.

Author: Good comment! With all these strange implementation that could very well happen. I'll add that.

       2. Else do:

>           1. If there are more than one NSEC record in the authority section then

Author: Fixed.

>              add name server IP to the *Multiple NSEC3* set and use one of them

Reviewer: [RFC 5155 § 7.2](https://datatracker.ietf.org/doc/html/rfc5155#section-7.2) states that responses may contain multiple NSEC3 RRs which “MUST use the same hash algorithm, iteration, and salt values”. But they can have different Flags field values. I don’t even know if there’s a way to deal with this situation that makes sense.

>              add name server IP to the *Multiple NSEC3* set and use one of them

Author: In this case we create a query that must be responded with a NODATA response (if the zone has NSEC3 and the zone name is apex). In that case there could only be one NSEC3 record.

>              add name server IP to the *Multiple NSEC3* set and use one of them

Author: > I think we should not be ambiguous here and just say to use the first one.

I missed the comment. I have updated to match your proposal.

    are non-empty then output *[DS03_SERVER_NO_NSEC3]* with the name server IP
    addresses from the *Responds Without NSEC3* set.

11. If the *Multiple NSEC3* set is non-empty then output *[DS03_ERR_MULT_NSEC3]*
    with the name server IP addresses from the set.

12. If the *Hash Algorithm* set is non-empty then do:
    1. If the set has more than one hash algorithm value then output
       *[DS03_INCONSISTENT_HASH_ALGO]*.
    2. For each algorithm value do:
       1. If the value is 1 output *[DS03_LEGAL_HASH_ALGO]* with the name servers
          IP addresses from the set with that value.
       2. Else, output *[DS03_ILLEGAL_HASH_ALGO]* with the hash algorithm value
          and the name servers IP addresses from the set with that value.

13. If the *NSEC3 Flags* set is non-empty then do:
    1. If the set has more than one flag list value then output
       *[DS03_INCONSISTENT_NSEC3_FLAGS]*.
    2. For each flag list value do:
       1. If any flag 0-6 (bits 0-6) is set then for each such flag output
          *[DS03_UNASSIGNED_FLAG_USED]* with the flag (bit) number and the name
          server IP addresses from the flag list value where the bit is set.

>        2. If flag 7 (bit 7) is set, then do:

Author: That is not what is meant.

       *[DS03_INCONSISTENT_SALT_LENTH]*.
    2. For each iteration value do:
       1. If the length is 0 output *[DS03_LEGAL_EMPTY_SALT]* with the name
          servers IP addresses from the set with that salt length.
       2. Else, output *[DS03_ILLEGAL_SALT_LENGTH]* with the length and the
          name servers IP addresses from the set with that salt lenght.

## Outcome(s)

The outcome of this Test Case is "fail" if there is at least one message
with the severity level *[ERROR]* or *[CRITICAL]*.

The outcome of this Test Case is "warning" if there is at least one message
with the severity level *[WARNING]*, but no message with severity level
*ERROR* or *CRITICAL*.

In other cases, no message or only messages with severity level
*[INFO]* or *[NOTICE]*, the outcome of this Test Case is "pass".

## Special procedural requirements

If either IPv4 or IPv6 transport is disabled, skip sending queries over that
transport protocol. A message will be outputted reporting that the transport
protocol has been skipped.

See the [DNSSEC README] document about DNSSEC algorithms.

## Intercase dependencies

None.

## Terminology

No special terminology for this Test Case.

> [Argument list]:                              https://github.com/zonemaster/zonemaster-engine/blob/master/docs/logentry_args.md

Author: Fixed.

[DS03_ERR_MULT_NSEC3]:                        #summary
[DS03_ILLEGAL_HASH_ALGO]:                     #summary
[DS03_ILLEGAL_ITERATION_VALUE]:               #summary
[DS03_ILLEGAL_SALT_LENGTH]:                   #summary
[DS03_INCONSISTENT_HASH_ALGO]:                #summary
[DS03_INCONSISTENT_ITERATION]:                #summary
[DS03_INCONSISTENT_NSEC3_FLAGS]:              #summary
[DS03_INCONSISTENT_SALT_LENTH]:               #summary
[DS03_LEGAL_EMPTY_SALT]:                      #summary
[DS03_LEGAL_HASH_ALGO]:                       #summary
[DS03_LEGAL_ITERATION_VALUE]:                 #summary
[DS03_NO_DNSSEC_SUPPORT]:                     #summary
[DS03_NO_NSEC3]:                              #summary
[DS03_NSEC3_OPT_OUT_DISABLED]:                #summary
[DS03_NSEC3_OPT_OUT_ENABLED_NON_TLD]:         #summary
[DS03_NSEC3_OPT_OUT_ENABLED_TLD]:             #summary
[DS03_SERVER_NO_DNSSEC_SUPPORT]:              #summary
[DS03_SERVER_NO_NSEC3]:                       #summary
[DS03_UNASSIGNED_FLAG_USED]:                  #summary
[ERROR]:                                      ../SeverityLevelDefinitions.md#error

> [IANA NSEC3 Parameterssregistry]:             https://www.iana.org/assignments/dnssec-nsec3-parameters/dnssec-nsec3-parameters.xhtml

Author: Fixed (three places).

## DNSSEC03: Verify NSEC3 parameters

### Test case identifier
**DNSSEC03**

## Table of contents

* [Objective](#objective)
* [Scope](#scope)
* [Inputs](#inputs)
* [Summary](#summary)
* [Test procedure]
* [Outcome(s)](#outcomes)
* [Special procedural requirements](#special-procedural-requirements)
* [Intercase dependencies](#intercase-dependencies)
* [Terminology](#terminology)

## Objective

The NSEC3 record type and its parameters are defined in [RFC 5155]. The
recommended values of the parameters have been updated by [RFC 9276].

For NSEC3 there are four fields that determine how the NSEC3 record are created

> and interpreted ([RFC 5155][RFC 5155#section-3], section3):

Reviewer: ```suggestion
and interpreted ([RFC 5155][RFC 5155#section-3], section 3):
```

Author: Fixed while editing the file.

* Hash algorithm
* Flags
* Iterations
* Salt

**Hash algorithm:** The only legal value of the hash algorithm is value 1
(SHA-1). See ([RFC 5155][RFC 5155#section-11], section 11 and
[IANA NSEC3 Parameters registry]).

> **Flags:** The only defined flags in the flag field is bit 7, "opt-out". It may

Reviewer: Depending on whom you ask, the least significant bit of a byte can be numbered 7 or 0. In RFC 5155, bit 7 is the least significant bit. It might be good to remind the reader of the convention that RFC 5155 chose.

Author: Added clarification. -- That is the way they always count in the RFC, isn't it?

Reviewer: You’re probably right. I never stopped to think about what the usual bit numbering convention is in network protocol specifications, actually. But it never hurts to remind the reader about those conventions.

At my old job, I spent a lot of time perusing old ITU-T recommendations about ISDN such as [Q.931](https://www.itu.int/rec/T-REC-Q.931-199805-I/en). In there, the convention is different: bits are numbered from 1 to 8, where bit 1 is the least-significant bit, and on ISDN networks, bits are transmitted least-significant bit first.

In bitwise arithmetic, it makes sense to number the bits from 0 to 7 where bit 0 is the least-significant bit: that way, bit N is worth 2^N if set. I helped @tgreenx a little bit with the implementation and we both fell into the trap of differing conventions.

DS03_NO_NSEC3                      | INFO    | ns_list | The zone does not use NSEC3. Testing for NSEC3 has been skipped. Fetched from name servers "{ns_list}".
DS03_NSEC3_OPT_OUT_DISABLED        | INFO    | ns_list | The following servers respond with NSEC3 opt-out disabled (as recommended). Fetched from name servers "{ns_list}".
DS03_NSEC3_OPT_OUT_ENABLED_NON_TLD | NOTICE  | ns_list | The following servers respond with NSEC3 opt-out enabled. The recommended practice is to disable opt-out. Fetched from name servers "{ns_list}".
DS03_NSEC3_OPT_OUT_ENABLED_TLD     | INFO    | ns_list | The following servers respond with NSEC3 opt-out enabled. Fetched from name servers "{ns_list}".
DS03_SERVER_NO_DNSSEC_SUPPORT      | ERROR   | ns_list | The following name servers do not support DNSSEC or have not been properly configured. Testing for NSEC3 has been skipped on those servers. Fetched from name servers "{ns_list}".
DS03_SERVER_NO_NSEC3               | ERROR   | ns_list | The following name servers do not use NSEC3, but others do. Testing for NSEC3 has been skipped on the following servers. Fetched from name servers "{ns_list}".

> DS03_UNASSIGNED_FLAG_USED          | ERROR   | ns_list, int | The following servers respond with an NSEC3 record where an unassigned flag is used (flag {int}). Fetched from name servers "{ns_list}".

Author: What about "(bit {int})"? The important message is that an unassigned flag is used. The exact flag in parenthesis.

> DS03_UNASSIGNED_FLAG_USED          | ERROR   | ns_list, int | The following servers respond with an NSEC3 record where an unassigned flag is used (flag {int}). Fetched from name servers "{ns_list}".

Reviewer: That looks fine to me.

          and the name servers IP addresses from the set with that value.

13. If the *NSEC3 Flags* set is non-empty then do:
    1. If the set has more than one flag list value then output
       *[DS03_INCONSISTENT_NSEC3_FLAGS]*.
    2. For each flag list value do:
       1. If any flag 0-6 (bits 0-6) is set then for each such flag output
          *[DS03_UNASSIGNED_FLAG_USED]* with the flag (bit) number and the name
          server IP addresses from the flag list value where the bit is set.
       2. If flag 7 (bit 7) is set, then do:
          1. If *Child Zone* is the root zone, a TLD zone or a zone matching
             *Public Suffix List Data* then output
             *[DS03_NSEC3_OPT_OUT_ENABLED_TLD]* with the name servers IP
             addresses from the set with that flag list value.
          2. Else, output *[DS03_NSEC3_OPT_OUT_ENABLED_NON_TLD]* with the name
             servers IP addresses from the set with that flag list value.
       3. If flag 7 (bit 7) is unset, then output
          *[DS03_NSEC3_OPT_OUT_DISABLED]* with the name servers IP addresses from
          the set with that flag list value.

> 13. If the *NSEC3 Iterations* set is non-empty then do:

Reviewer: ```suggestion
14. If the *NSEC3 Iterations* set is non-empty then do:
```

Author: Fixed while updating.

    1. If the set has more than one iteration value then output
       *[DS03_INCONSISTENT_ITERATION]*.
    2. For each iteration value do:
       1. If the value is 0 output *[DS03_LEGAL_ITERATION_VALUE]* with the name
          servers IP addresses from the set with that iteration value.
       2. Else, output *[DS03_ILLEGAL_ITERATION_VALUE]* with the value and the
          name servers IP addresses from the set with that iteration value.

> 13. If the *NSEC3 Salt Length* set is non-empty then do:

Reviewer: ```suggestion
15. If the *NSEC3 Salt Length* set is non-empty then do:
```

Author: Fixed while updating.

[Public Suffix List data]:                    https://publicsuffix.org/list/public_suffix_list.dat
[Public Suffix List]:                         https://publicsuffix.org/list/
[RCODE Name]:                                 https://www.iana.org/assignments/dns-parameters/dns-parameters.xhtml#dns-parameters-6
[RFC 5155#section-11]:                        https://www.rfc-editor.org/rfc/rfc5155.html#section-11
[RFC 5155#section-3]:                         https://www.rfc-editor.org/rfc/rfc5155.html#section-3
[RFC 5155]:                                   https://www.rfc-editor.org/rfc/rfc5155.html
[RFC 9276#section-2.4]:                       https://www.rfc-editor.org/rfc/rfc9276.html#section-2.4
[RFC 9276#section-3.1]:                       https://www.rfc-editor.org/rfc/rfc9276.html#section-3.1
[RFC 9276]:                                   https://www.rfc-editor.org/rfc/rfc9276.html
[Severity Level Definitions]:                 ../SeverityLevelDefinitions.md
[Test procedure]:                             #test-procedure
[WARNING]:                                    ../SeverityLevelDefinitions.md#warning
[Zonemaster-Engine profile]:                  ../../../configuration/profiles.md

> B

Author: Fixed.

       matching *Child Zone* in the answer section, add name server IP to the
       *Responds Without DNSKEY* set and go to next name server.
    4. Add name server IP to the *Responds With DNSKEY* set.
    5. Send *NSEC Query* to the name server IP.
    6. If there is no DNS response do:
       1. Add name server IP to the *No Response NSEC Query* set.
       2. Go to next name server IP.
    7. If the [RCODE Name] in the response is not "NoError" or if the AA flag is
       not set in the response (or both) then do:
       1. Add name server IP to the *Error Response NSEC Query* set.
       2. Go to next name server IP.
    8. If the authority section contains no NSEC3 record then add the name server
       IP to the *Responds Without NSEC3* set and go to next name server.
    9. Else do:

>        1. If there are more than one NSECS record in the authority section then

Reviewer: There’s a typo: “NSECS”. Shouldn’t it be “NSEC3”?
